### PR TITLE
Add hover for classes, modules and constants

### LINKED ISSUE #201: Add hover to the `ruby-lsp`
If https://github.com/Shopify/ruby-lsp/issues/199 works and is successful in providing decent results for finding possible method definitions, we can use the same mechanism to provide hover. This request shows documentation/information about methods and constant when hovering with the cursor.

Note: there may be conflicts between possible methods and constants. It's likely that we'll only be able to provide hover if only a single result is found in the index.

Documentation: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_hover

### PR DESCRIPTION
**Pull request by vinistock**
### Motivation

Closes #201

Add hover for classes, modules and constants based on what's indexed. Because Sorbet can fully handle displaying documentation on hover even for gems, I think we can completely skip the functionality if Sorbet is present.

Alternatively, we could instead drop the functionality from Tapioca to collect comments in RBIs and enable this to work like definition. I'd appreciate thoughts on this one.

### Implementation

I highly recommend reviewing by commit:
1. Removed all files related to the Rails documentation feature
2. Remove railties (only related to the Rails documentation feature)
3. Allowed passing fragments to `URI.from_path`, which is necessary for building definition links
4. Added hover based on the index

### What about the Rails documentation feature?

All that functionality will be moved to `ruby-lsp-rails`. I'm marking this PR as a breaking change so that we can easily coordinate moving the feature there, by bumping the `ruby-lsp-rails` requirement only when it's moved.

### Automated Tests

Added an example for each type of indexed symbol.

### CODE REVIEW

File: lib/ruby_lsp/executor.rb [hunk 1]
```diff
@@ -252,16 +252,20 @@ def hover(uri, position)
...
+      target, parent, nesting = document.locate_node(
+        position,
+        node_types: Requests::Hover::ALLOWED_TARGETS,
+      )
 
-      if !Requests::Hover::ALLOWED_TARGETS.include?(target.class) &&
-          Requests::Hover::ALLOWED_TARGETS.include?(parent.class)
+      if (Requests::Hover::ALLOWED_TARGETS.include?(parent.class) &&
+          !Requests::Hover::ALLOWED_TARGETS.include?(target.class)) ||
+          (parent.is_a?(SyntaxTree::ConstPathRef) && target.is_a?(SyntaxTree::Const))
```
[the remark this answers is not in the document]
Comment (Morriar) [reply lib/ruby_lsp/executor.rb:269]: Should we put an issue somewhere to not forget to do this cleaning once everything is migrated?

File: lib/ruby_lsp/requests/hover.rb [hunk 2]
```diff
@@ -61,34 +67,52 @@ def merge_response!(other)
...
+      def find_in_index(name, node)
+        entries = @index.resolve(name, @nesting)
+        return unless entries
+
+        title = +"```ruby\n#{name}\n```"
+        definitions = []
+        content = +""
+        entries.each do |entry|
+          loc = entry.location
+          file_name = Pathname.new(entry.file_path).basename
```
Comment (st0012) [inline lib/ruby_lsp/requests/hover.rb:97]: I wonder if this should be a method of `Entry` 🤔

File: test/expectations/hover/documented_module.exp.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    "params": [
+        {
+            "line": 1,
+            "character": 7
+        }
+    ],
+    "result": {
+        "contents": {
+            "kind": "markdown",
+            "value": "```ruby\nFoo\n```\n\n**Definitions**: [fake.rb](file:///fake.rb#L2,1-3,3)\n\n\n\nThis is the documentation for Foo"
```
Comment (Morriar) [inline test/expectations/hover/documented_module.exp.json:11]: Can we add a test with multiple entries for the same name?

File: lib/ruby_lsp/requests/hover.rb (hunk 2 above)
Comment (vinistock) [reply lib/ruby_lsp/requests/hover.rb]: Good idea!

File: lib/ruby_lsp/requests/hover.rb [hunk 4]
```diff
@@ -61,34 +67,52 @@ def merge_response!(other)
...
       end
 
       private
 
-      sig { params(name: String, node: SyntaxTree::Node).returns(T.nilable(Interface::Hover)) }
-      def generate_rails_document_link_hover(name, node)
-        urls = Support::RailsDocumentClient.generate_rails_document_urls(name)
-        return if urls.empty?
+      sig { params(name: String, node: SyntaxTree::Node).void }
+      def find_in_index(name, node)
```
[the remark this answers is not in the document]
Comment (vinistock) [reply lib/ruby_lsp/requests/hover.rb]: Changed it!

**Comment by vinistock** [comment]
I haven't moved the old hover functionality to `ruby-lsp-rails` yet. Once it's complete, I'll share what it looks like with the two hovers.

File: lib/ruby_lsp/requests/hover.rb [hunk 5]
```diff
@@ -61,34 +67,51 @@ def merge_response!(other)
...
+
+          # We always handle locations as zero based. However, for file links in Markdown we need them to be one based,
+          # which is why instead of the usual subtraction of 1 to line numbers, we are actually adding 1 to columns. The
+          # format for VS Code file URIs is `file:///path/to/file.rb#Lstart_line,start_column-end_line,end_column`
+          uri = URI::Generic.from_path(
+            path: entry.file_path,
+            fragment: "L#{loc.start_line},#{loc.start_column + 1}-#{loc.end_line},#{loc.end_column + 1}",
+          )
+
+          definitions << "[#{entry.file_name}](#{uri})"
```
Comment (Morriar) [inline lib/ruby_lsp/requests/hover.rb:106]: Are paths too long to be displayed here?

**Review by Morriar — APPROVED** [review]

**Pull request merged**